### PR TITLE
Semaphores and condition variables.

### PR DESCRIPTION
**Pull request by Apprentice-Alchemist**
- [x] hxcpp (HaxeFoundation/hxcpp#976)
- [x] C#
- [x] Java/Jvm
- [x] Python
- [x] Eval
- [x] Hashlink (HaxeFoundation/hashlink#495)

Since neko is deprecated I have not implemented these for neko.

### CODE REVIEW

**Comment by RealyUniqueName** [comment]
Nice!
Pls, fix building so we can check CI.

File: opam [hunk 1]
```diff
@@ -19,7 +19,7 @@ build: [
 install: [make "install" "INSTALL_DIR=%{prefix}%"]
 remove: [make "uninstall" "INSTALL_DIR=%{prefix}%"]
 depends: [
-  "ocaml" {>= "4.02"}
+  "ocaml" {>= "4.12"}
```
Comment (RealyUniqueName) [inline opam]: @andyli do we still have to stick to 4.02?

Comment (andyli) [reply opam]: Ubuntu 16.04 reached its [End of Standard Support](https://wiki.ubuntu.com/Releases) earlier this year. We can bump the min ocaml version to 4.05.0, which is provided by Ubuntu 18.04.

**Comment by Apprentice-Alchemist** [comment]
> Please don't do this, there are some implications related to packaging that we would like to avoid.

Changes to the ocaml version have been reverted.
Using luv to implement semaphores and condition variables seems to work.

**Comment by RealyUniqueName** [comment]
I think luv's synchronization primitives require explicit calls to `destroy` methods for cleanups.

File: src/macro/eval/evalHash.ml [hunk 2]
```diff
@@ -141,7 +141,8 @@ let key_sys_net_Lock = hash "sys.thread.Lock"
 let key_sys_net_Tls = hash "sys.thread.Tls"
 let key_sys_net_Deque = hash "sys.thread.Deque"
 let key_sys_thread_EventLoop = hash "sys.thread.EventLoop"
-
+(* let key_sys_thread_Semaphore = hash "sys.thread.Semaphore"
```
Comment (RealyUniqueName) [inline src/macro/eval/evalHash.ml]: These changes are no longer needed. Please revert ml files.

File: std/cs/_std/sys/thread/Semaphore.hx [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+	public function release():Void {
+		native.Release();
+	}
+}
+
+// doesn't seem to be in the dlls shipped with hxcs?
+
+@:native("System.Threading.SemaphoreSlim")
+private extern class SemaphoreSlim implements cs.system.IDisposable {
```
Comment (RealyUniqueName) [inline std/cs/_std/sys/thread/Semaphore.hx]: SemaphoreSlim was added in .NET 4.0. Our Haxe targets 2.0 by default.

Comment (Apprentice-Alchemist) [reply std/cs/_std/sys/thread/Semaphore.hx]: It uses `System.Threading.Semaphore` now.

**Comment by ncannasse** [comment]
I'm curious about the efficiency of these VS implementing them on top of Lock/Mutex :)

**Comment by Apprentice-Alchemist** [comment]
Lock is already implemented using a semaphore on many targets.
So to implement a semaphore, you'd end up using a semaphore + an extra counter + a mutex to protect that counter.
That sounds less efficient that using a semaphore directly.

**Comment by Simn** [comment]
Sorry for the delay! I've merged the hxcpp PR, could you update this one so CI runs?

Edit: Uhm, wrong PR, sorry!

**Comment by Simn** [comment]
Now my last comment is accurate, so could you push again? I still don't know why I can't restart CI on this one.

**Comment by Apprentice-Alchemist** [comment]
`windows64-test (cpp)` failure: Turns out hxcpp already sets the `HXCPP_WINXP_COMPAT` by default.
Simplest fix (in my opinion) is to turn that off: https://github.com/HaxeFoundation/hxcpp/pull/984
`windows-build` failure: no idea, it's not showing any logs.

**Comment by Simn** [comment]
Had to restart CI again because `CI / linux-arm64 (pull_request)` was hanging. Let's see if this finally turns green now...

**Comment by Simn** [comment]
Whatever, I'll just merge this!

**Pull request merged**